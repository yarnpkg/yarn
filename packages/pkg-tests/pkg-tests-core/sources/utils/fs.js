/* @flow */

import type {Gzip} from 'zlib';

const fs = require('fs-extra');
const klaw = require('klaw');
const path = require('path');
const tarFs = require('tar-fs');
const tmp = require('tmp');
const zlib = require('zlib');

const miscUtils = require('./misc');

exports.walk = function walk(
  source: string,
  {filter, relative = false}: {|filter?: Array<string>, relative?: boolean|} = {},
): Promise<Array<string>> {
  return new Promise((resolve, reject) => {
    const paths = [];

    const walker = klaw(source, {
      filter: itemPath => {
        if (!filter) {
          return true;
        }

        const stat = fs.statSync(itemPath);

        if (stat.isDirectory()) {
          return true;
        }

        const relativePath = path.relative(source, itemPath);

        if (miscUtils.filePatternMatch(relativePath, filter)) {
          return true;
        }

        return false;
      },
    });

    walker.on('data', ({path: itemPath}) => {
      const relativePath = path.relative(source, itemPath);

      if (!filter || miscUtils.filePatternMatch(relativePath, filter)) {
        paths.push(relative ? relativePath : itemPath);
      }

      // This item has been accepted only because it's a directory; it doesn't match the filter
      return;
    });

    walker.on('end', () => {
      resolve(paths);
    });
  });
};

exports.packToStream = function packToStream(
  source: string,
  // $FlowFixMe Flow doesn't support null default parameters in exact types
  {virtualPath = null}: {|virtualPath?: ?string|} = {},
): Gzip {
  if (virtualPath) {
    if (!path.isAbsolute(virtualPath)) {
      throw new Error('The virtual path has to be an absolute path');
    } else {
      virtualPath = path.resolve(virtualPath);
    }
  }

  const zipperStream = zlib.createGzip();

  const packStream = tarFs.pack(source, {
    map: header => {
      if (true) {
        header.name = path.resolve('/', header.name);
        header.name = path.relative('/', header.name);
      }

      if (virtualPath) {
        header.name = path.resolve('/', virtualPath, header.name);
        header.name = path.relative('/', header.name);
      }

      return header;
    },
  });

  packStream.pipe(zipperStream);

  packStream.on('error', error => {
    zipperStream.emit('error', error);
  });

  return zipperStream;
};

exports.packToFile = function packToFile(target: string, source: string, options: *): Promise<void> {
  const tarballStream = fs.createWriteStream(target);

  const packStream = exports.packToStream(source, options);
  packStream.pipe(tarballStream);

  return new Promise((resolve, reject) => {
    tarballStream.on('error', error => {
      reject(error);
    });

    packStream.on('error', error => {
      reject(error);
    });

    tarballStream.on('close', () => {
      resolve();
    });
  });
};

exports.createTemporaryFolder = function createTemporaryFolder(): Promise<string> {
  return new Promise((resolve, reject) => {
    tmp.dir({unsafeCleanup: true}, (error, dirPath) => {
      if (error) {
        reject(error);
      } else {
        resolve(dirPath);
      }
    });
  });
};

exports.createTemporaryFile = async function createTemporaryFile(filePath: string): Promise<string> {
  if (filePath) {
    if (path.normalize(filePath).match(/^(\.\.)?\//)) {
      throw new Error('A temporary file path must be a forward path');
    }

    const folderPath = await exports.createTemporaryFolder();
    return path.resolve(folderPath, filePath);
  } else {
    return new Promise((resolve, reject) => {
      tmp.file({discardDescriptor: true}, (error, filePath) => {
        if (error) {
          reject(error);
        } else {
          resolve(filePath);
        }
      });
    });
  }
};

exports.writeFile = async function writeFile(target: string, body: string | Buffer): Promise<void> {
  await fs.mkdirp(path.dirname(target));
  await fs.writeFile(target, body);
};

exports.readFile = function readFile(source: string, encoding: ?string = null): Promise<any> {
  return fs.readFile(source, encoding);
};

exports.writeJson = function writeJson(target: string, object: any): Promise<void> {
  return exports.writeFile(target, JSON.stringify(object));
};

exports.readJson = async function readJson(source: string): Promise<any> {
  const fileContent = await exports.readFile(source, 'utf8');

  try {
    return JSON.parse(fileContent);
  } catch (error) {
    throw new Error(`Invalid json file (${source})`);
  }
};
