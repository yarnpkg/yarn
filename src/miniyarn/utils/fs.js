import * as Fse from 'fs-extra';
import * as Fs from 'fs';
import gunzipMaybe from 'gunzip-maybe';
import klaw from 'klaw';
import Path from 'path';
import rcopy from 'recursive-copy';
import tarFs from 'tar-fs';
import tar from 'tar-stream';
import tmp from 'tmp';
import {createGzip} from 'zlib';

import * as miscUtils from 'miniyarn/utils/misc';
import * as pathUtils from 'miniyarn/utils/path';
import * as parseUtils from 'miniyarn/utils/parse';

let isWindows = null;

export async function detectWindows() {
  if (isWindows !== null) return isWindows;

  if (await exists(`c:/`)) return (isWindows = true);

  if ((await exists(`/proc/version`)) && (await readFile(`/proc/version`, `utf8`)).match(/Microsoft|WSL/))
    return (isWindows = true);

  return (isWindows = false);
}

export async function cwd(path, callback) {
  let cwd = process.cwd();
  process.chdir(path);

  try {
    return await callback();
  } finally {
    process.chdir(cwd);
  }
}

export async function symlink(source, destination) {
  return await new Promise((resolve, reject) => {
    Fs.symlink(source, destination, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function ensureSymlink(source, destination) {
  return await new Promise((resolve, reject) => {
    Fse.ensureSymlink(source, destination, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function isDirectory(path) {
  return await new Promise((resolve, reject) => {
    Fs.lstat(path, (error, stat) => {
      if (error) {
        reject(error);
      } else {
        resolve(stat.isDirectory());
      }
    });
  });
}

export async function createTemporaryFolder() {
  return new Promise((resolve, reject) => {
    tmp.dir({unsafeCleanup: true}, (error, path) => {
      if (error) {
        reject(error);
      } else {
        resolve(path);
      }
    });
  });
}

export async function createTemporaryFile(filePath) {
  if (filePath) {
    if (!pathUtils.isForward(filePath)) throw new Error(`A temporary file path must be a forward path`);

    let folderPath = await createTemporaryFolder();

    return pathUtils.resolve(folderPath, filePath);
  } else {
    return await new Promise((resolve, reject) => {
      tmp.file((error, path) => {
        if (error) {
          reject(error);
        } else {
          resolve(path);
        }
      });
    });
  }
}

export async function walk(path, {filter} = {}) {
  return await new Promise((resolve, reject) => {
    let paths = [];

    let walker = klaw(path, {
      filter: itemPath => {
        if (!filter) return true;

        if (Fse.statSync(path).isDirectory()) return true;
        s;
        let relativePath = pathUtils.relative(path, itemPath);

        if (miscUtils.filePatternMatch(relativePath, filter)) return true;

        return false;
      },
    });

    walker.on(`data`, ({path: itemPath}) => {
      if (!filter) return void paths.push(itemPath);

      let relativePath = pathUtils.relative(path, itemPath);

      if (miscUtils.filePatternMatch(relativePath, filter)) return void paths.push(itemPath);

      return; // This item has been accepted only because it's a directory; it doesn't match the filter
    });

    walker.on(`end`, () => {
      resolve(paths);
    });
  });
}

export async function mkdirp(path) {
  return await new Promise((resolve, reject) => {
    Fse.ensureDir(path, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function rm(path) {
  if (path === ``) throw new Error(`Cannot rm an empty path`);

  if (path === `/`) throw new Error(`Cannot rm your whole / directory - something's very wrong`);

  if (!await exists(path)) return;

  return await new Promise((resolve, reject) => {
    Fse.remove(path, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export async function getMTime(path) {
  return await new Promise((resolve, reject) => {
    Fs.lstat(path, (error, stat) => {
      if (error) {
        reject(error);
      } else {
        resolve(stat.mtime);
      }
    });
  });
}

export async function createFileWriter(path) {
  await mkdirp(pathUtils.dirname(path));

  let stream = Fs.createWriteStream(path);

  stream.promise = new Promise((resolve, reject) => {
    stream.on(`finish`, () => {
      resolve();
    });

    stream.on(`error`, error => {
      reject(error);
    });
  });

  return stream;
}

export function createFileReader(path) {
  let stream = Fs.createReadStream(path);

  stream.promise = new Promise((resolve, reject) => {
    stream.on(`error`, error => {
      reject(error);
    });

    stream.on(`end`, () => {
      resolve();
    });
  });

  return stream;
}

export async function exists(path) {
  return await new Promise((resolve, reject) => {
    Fs.exists(path, status => {
      resolve(status);
    });
  });
}

export async function cp(source, destination, {filter = []} = {}) {
  await mkdirp(pathUtils.dirname(destination));
  await rcopy(source, destination, {dot: true, filter: path => miscUtils.filePatternMatch(path, filter)});
}

export async function mv(source, target) {
  // Windows defender often lock files after they are created, which makes it impossible to move them.
  // We could have a mechanism of "try again", but it seems simpler to just use a copy operation instead.
  // Note that using cp instead of mv is not strictly comparable since, unfortunately, it's not atomic. Nothing we can do, tho :(

  if (await detectWindows()) {
    await cp(source, target);
    await rm(source);
  } else {
    await mkdirp(pathUtils.dirname(target));

    return await new Promise((resolve, reject) => {
      Fs.rename(source, target, err => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
  }
}

export async function extract(target, tarballPath, {stripLeadingFolder = false} = {}) {
  return await new Promise((resolve, reject) => {
    function map(header) {
      if (header.name[0] === `/`) header.name = header.name.replace(/^\/+/, ``);

      if (stripLeadingFolder) header.name = header.name.replace(/^[^\/]+\//, ``);

      return header;
    }

    let unzipper = gunzipMaybe();
    Fs.createReadStream(tarballPath).pipe(unzipper);

    let extractor = tarFs.extract(target, {map});
    unzipper.pipe(extractor);

    extractor.on(`error`, error => {
      reject(error);
    });

    extractor.on(`finish`, () => {
      resolve();
    });
  });
}

export function packToStream(path, {virtualPath = null} = {}) {
  if (virtualPath) {
    if (!pathUtils.isAbsolute(virtualPath)) {
      throw new Error(`The virtual path has to be an absolute path`);
    } else {
      virtualPath = pathUtils.resolve(virtualPath);
    }
  }

  let zipperStream = createGzip();

  let packStream = tarFs.pack(path, {
    map: header => {
      if (true) {
        header.name = pathUtils.resolve(`/`, header.name);
        header.name = pathUtils.relative(`/`, header.name);
      }

      if (virtualPath) {
        header.name = pathUtils.resolve(`/`, virtualPath, header.name);
        header.name = pathUtils.relative(`/`, header.name);
      }

      return header;
    },
  });

  packStream.pipe(zipperStream);

  packStream.on(`error`, error => {
    zipperStream.emit(`error`, error);
  });

  return zipperStream;
}

export async function packToFile(target, directoryPath, {...options} = {}) {
  let tarballStream = Fs.createWriteStream(target);

  let packStream = packToStream(directoryPath, options);
  packStream.pipe(tarballStream);

  return await new Promise((resolve, reject) => {
    tarballStream.on(`error`, error => {
      reject(error);
    });

    packStream.on(`error`, error => {
      reject(error);
    });

    tarballStream.on(`close`, () => {
      resolve();
    });
  });
}

export async function writeFile(path, body) {
  await mkdirp(pathUtils.dirname(path));

  return await new Promise((resolve, reject) => {
    Fs.writeFile(path, body, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export async function readFile(path, encoding = null) {
  return await new Promise((resolve, reject) => {
    Fs.readFile(path, encoding, (err, body) => {
      if (err) {
        reject(err);
      } else {
        resolve(body);
      }
    });
  });
}

export async function readJson(path) {
  let fileContent = await readFile(path, `utf8`);

  return parseUtils.parseJson(fileContent);
}

export async function readDirectory(path) {
  return await new Promise((resolve, reject) => {
    Fs.readdir(path, (err, entries) => {
      if (err) {
        reject(err);
      } else {
        resolve(entries);
      }
    });
  });
}

export class Handler {
  constructor(path, {temporary = false} = {}) {
    this.path = path;

    this.temporary = temporary;
  }

  get() {
    return this.path;
  }

  async steal() {
    if (this.temporary) return this.path;

    let destination = (await isDirectory(this.path)) ? await createTemporaryFolder() : await createTemporaryFile();

    await cp(this.path, destination);

    return destination;
  }
}
