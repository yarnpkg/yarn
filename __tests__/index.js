/* @flow */

import NoopReporter from '../src/reporters/base-reporter.js';
import makeTemp from './_temp';
import * as fs from '../src/util/fs.js';
const pkg = require('../package.json');

const path = require('path');
const exec = require('child_process').exec;

const fixturesLoc = path.join(__dirname, './fixtures/index');
const yarnBin = path.join(__dirname, '../bin/yarn.js');
const semver = require('semver');
let ver = process.versions.node;
ver = ver.split('-')[0];

jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

async function execCommand(cmd: string, args: Array<string>, name: string, makeTempDir: ?boolean):
Promise<Array<?string>> {
  const srcDir = path.join(fixturesLoc, name);
  let workingDir = srcDir;
  if (makeTempDir) {
    workingDir = await makeTemp(name);
    await fs.copy(srcDir, workingDir, new NoopReporter());
  }

  return new Promise((resolve, reject) => {
    exec(`node "${yarnBin}" ${cmd} ${args.join(' ')}`, {cwd:workingDir, env:process.env}, (error, stdout) => {
      if (error) {
        reject({error, stdout});
      } else {
        const stdoutLines = stdout.toString()
          .split('\n')
          .map((line: ?string) => line && line.trim())
          .filter((line: ?string) => line);

        resolve(stdoutLines);
      }
    });
  });
}

function expectAddSuccessfullOutput(stdout, pkg) {
  const lastLines = stdout.slice(stdout.length - 4);
  expect(lastLines[0]).toEqual('success Saved lockfile.');
  expect(lastLines[1]).toEqual('success Saved 1 new dependency.');
  expect(lastLines[2]).toContain(pkg);
  expect(lastLines[3]).toContain('Done');
}

function expectAddSuccessfullOutputWithNoLockFile(stdout, pkg) {
  const lastLines = stdout.slice(stdout.length - 4);
  expect(lastLines[1]).toEqual('success Saved 1 new dependency.');
  expect(lastLines[2]).toContain(pkg);
  expect(lastLines[3]).toContain('Done');
}

function expectRunOutput(stdout) {
  const lastLines = stdout.slice(stdout.length - 2);
  expect(lastLines[0]).toMatch(/A message from custom script/);
  expect(lastLines[1]).toMatch(/^Done/);
}

function expectHelpOutput(stdout) {
  expect(stdout[0]).toEqual('Usage: yarn [command] [flags]');
  const lastLines = stdout.slice(stdout.length - 2);
  expect(lastLines[0]).toEqual('Run `yarn help COMMAND` for more information on specific commands.');
  expect(lastLines[1]).toEqual('Visit https://yarnpkg.com/en/docs/cli/ to learn more about Yarn.');
}

function expectHelpOutputAsSubcommand(stdout) {
  expect(stdout[0]).toEqual('Usage: yarn add [packages ...] [flags]');
  expect(stdout[stdout.length - 1])
    .toEqual('Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.');
}

function expectAnErrorMessage(command: Promise<Array<?string>>, error: string) : Promise<void> {
  return command
  .then(function() {
    throw new Error('the command did not fail');
  })
  .catch((reason) =>
    expect(reason.error.message).toContain(error),
  );
}

function expectAnInfoMessageAfterError(command: Promise<Array<?string>>, info: string) : Promise<void> {
  return command
  .then(function() {
    throw new Error('the command did not fail');
  })
  .catch((reason) =>
    expect(reason.stdout).toContain(info),
  );
}

test.concurrent('should add package', async () => {
  const stdout = await execCommand('add', ['left-pad'], 'run-add', true);
  expectAddSuccessfullOutput(stdout, 'left-pad');
});

test.concurrent('should add package with no-lockfile option', async () => {
  const stdout = await execCommand('add', ['repeating', '--no-lockfile'], 'run-add-option', true);
  expectAddSuccessfullOutputWithNoLockFile(stdout, 'repeating');
});

test.concurrent('should add package with no-lockfile option in front', async () => {
  const stdout = await execCommand('add', ['--no-lockfile', 'split-lines'], 'run-add-option-in-front', true);
  expectAddSuccessfullOutputWithNoLockFile(stdout, 'split-lines');
});

test.concurrent('should add lockfile package', async () => {
  const stdout = await execCommand('add', ['lockfile'], 'run-add-lockfile', true);
  expectAddSuccessfullOutput(stdout, 'lockfile');
});

// test is failing on Node 4, https://travis-ci.org/yarnpkg/yarn/jobs/216254539
if (semver.satisfies(ver, '>=5.0.0')) {
  test.concurrent('should add progress package globally', async () => {
    const stdout = await execCommand('global',
      ['add', 'progress', '--global-folder', './global'],
      'run-add-progress-globally',
      true);

    const lastLine = stdout[stdout.length - 1];
    expect(lastLine).toMatch(/^Done/);
  });
}

test.concurrent('should run custom script', async () => {
  const stdout = await execCommand('run', ['custom-script'], 'run-custom-script');
  expectRunOutput(stdout);
});

test.concurrent('should run custom script without run command', async () => {
  const stdout = await execCommand('custom-script', [], 'run-custom-script');
  expectRunOutput(stdout);
});

test.concurrent('should run help command', async () => {
  const stdout = await execCommand('help', [], 'run-help');
  expectHelpOutput(stdout);
});

test.concurrent('should run help command with --help', async () => {
  const stdout = await execCommand('--help', [], 'run-help');
  expectHelpOutput(stdout);
});

test.concurrent('should run help command with -h', async () => {
  const stdout = await execCommand('-h', [], 'run-help');
  expectHelpOutput(stdout);
});

test.concurrent('should run add command with help option', async () => {
  const stdout = await execCommand('add', ['--help'], 'run-help');
  expectHelpOutputAsSubcommand(stdout);
});

test.concurrent('should run add command with h option', async () => {
  const stdout = await execCommand('add', ['-h'], 'run-help');
  expectHelpOutputAsSubcommand(stdout);
});

test.concurrent('should run help command with add option', async () => {
  const stdout = await execCommand('help', ['add'], 'run-help');
  expectHelpOutputAsSubcommand(stdout);
});

test.concurrent('should run --help command with add option', async () => {
  const stdout = await execCommand('--help', ['add'], 'run-help');
  expectHelpOutputAsSubcommand(stdout);
});

test.concurrent('should run -h command with add option', async () => {
  const stdout = await execCommand('-h', ['add'], 'run-help');
  expectHelpOutputAsSubcommand(stdout);
});

test.concurrent('should run version command', async () => {
  await expectAnErrorMessage(
    execCommand('version', [], 'run-version'),
    'Can\'t answer a question unless a user TTY',
  );
});

test.concurrent('should run --version command', async () => {
  const stdout = await execCommand('--version', [], 'run-version');
  expect(stdout[0]).toEqual(pkg.version);
});

test.concurrent('should install if no args', async () => {
  const stdout = await execCommand('', [], 'run-add', true);
  expect(stdout[0]).toEqual(`yarn install v${pkg.version}`);
});

test.concurrent('should install if first arg looks like a flag', async () => {
  const stdout = await execCommand('--json', [], 'run-add', true);
  expect(stdout[stdout.length - 1]).toEqual('{"type":"success","data":"Saved lockfile."}');
});

test.concurrent('should interpolate aliases', async () => {
  await expectAnErrorMessage(
    execCommand('i', [], 'run-add', true),
    'Did you mean `yarn install`?',
  );
});

test.concurrent('should display correct documentation link for aliases', async () => {
  await expectAnInfoMessageAfterError(
    execCommand('i', [], 'run-add', true),
    'Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.',
  );
});

test.concurrent('should run help of run command if --help is before --', async () => {
  const stdout = await execCommand('run', ['custom-script', '--help', '--'], 'run-custom-script-with-arguments');
  expect(stdout[0]).toEqual('Usage: yarn [command] [flags]');
  expect(stdout[stdout.length - 1])
    .toEqual('Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.');
});

if (process.platform !== 'win32') {
  test.concurrent('should run help of custom-script if --help is after --', async () => {
    const stdout = await execCommand('run', ['custom-script', '--', '--help'], 'run-custom-script-with-arguments');
    expect(stdout[stdout.length - 2]).toEqual('A message from custom script with args --help');
  });
}

test.concurrent('should run bin command', async () => {
  const stdout = await execCommand('bin', [], '', false);
  expect(stdout[0]).toEqual(path.join(fixturesLoc, 'node_modules', '.bin'));
  expect(stdout.length).toEqual(1);
});

test.concurrent('should throws missing command for not camelised command', async () => {
  await expectAnErrorMessage(
    execCommand('HelP', [], 'run-add', true),
    'Command \"HelP\" not found',
  );
});

test.concurrent('should throws missing command for not alphabetic command', async () => {
  await expectAnErrorMessage(
    execCommand('123', [], 'run-add', true),
    'Command \"123\" not found',
  );
});

test.concurrent('should throws missing command for unknown command', async () => {
  await expectAnErrorMessage(
    execCommand('unknown', [], 'run-add', true),
    'Command \"unknown\" not found',
  );
});

test.concurrent('should not display documentation link for unknown command', async () => {
  await expectAnInfoMessageAfterError(
    execCommand('unknown', [], 'run-add', true),
    '',
  );
});

test.concurrent('should display documentation link for known command', async () => {
  await expectAnInfoMessageAfterError(
    execCommand('add', [], 'run-add', true),
    'Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.',
  );
});

test.concurrent('should throws missing command for constructor command', async () => {
  await expectAnErrorMessage(
    execCommand('constructor', [], 'run-add', true),
    'Command \"constructor\" not found',
  );
});
