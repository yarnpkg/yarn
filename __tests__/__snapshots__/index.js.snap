// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add lockfile package 1`] = `
Array [
  "info No lockfile found.",
  "[1/4] Resolving packages...",
  "[2/4] Fetching packages...",
  "[3/4] Linking dependencies...",
  "[4/4] Building fresh packages...",
  "success Saved lockfile.",
  "success Saved 1 new dependency.",
  "info Direct dependencies",
  "└─ lockfile@1.0.3",
  "info All dependencies",
  "└─ lockfile@1.0.3",
]
`;

exports[`should add package 1`] = `
Array [
  "info No lockfile found.",
  "[1/4] Resolving packages...",
  "[2/4] Fetching packages...",
  "[3/4] Linking dependencies...",
  "[4/4] Building fresh packages...",
  "success Saved lockfile.",
  "success Saved 1 new dependency.",
  "info Direct dependencies",
  "└─ left-pad@1.2.0",
  "info All dependencies",
  "└─ left-pad@1.2.0",
]
`;

exports[`should add package with frozen-lockfile option 1`] = `
Array [
  "info No lockfile found.",
  "[1/4] Resolving packages...",
  "[2/4] Fetching packages...",
  "[3/4] Linking dependencies...",
  "[4/4] Building fresh packages...",
  "success Saved 1 new dependency.",
  "info Direct dependencies",
  "└─ repeating@3.0.0",
  "info All dependencies",
  "└─ repeating@3.0.0",
]
`;

exports[`should add package with no-lockfile option 1`] = `
Array [
  "info No lockfile found.",
  "[1/4] Resolving packages...",
  "[2/4] Fetching packages...",
  "[3/4] Linking dependencies...",
  "[4/4] Building fresh packages...",
  "success Saved 1 new dependency.",
  "info Direct dependencies",
  "└─ repeating@3.0.0",
  "info All dependencies",
  "└─ repeating@3.0.0",
]
`;

exports[`should add package with no-lockfile option in front 1`] = `
Array [
  "info No lockfile found.",
  "[1/4] Resolving packages...",
  "[2/4] Fetching packages...",
  "[3/4] Linking dependencies...",
  "[4/4] Building fresh packages...",
  "success Saved 1 new dependency.",
  "info Direct dependencies",
  "└─ split-lines@1.1.0",
  "info All dependencies",
  "└─ split-lines@1.1.0",
]
`;
